### ISSUE TITLE
Few suggestions - generic relationship schema, automatic follow apps type for CRD and more...

### ISSUE THREAD

**Issue by shay-berman** [cloud-ark/kubeplus#627]
@devdattakulkarni I saw your session in kubecon2020 NA about operators, so I decided to try the plugin on [RedisEnterpriseCluster](https://github.com/RedisLabs/redis-enterprise-k8s-docs).

Opened this ticket to give you some suggestions on improving the plugin:

**Suggestions list**:
1. Since most of CR creates an app (type of sts, deplp or daemonset) I would suggest you that by design when you run kubectl connections CRDx objecty you should look also on these apps types by default.

2. Adding annotations is good way to indicate for relationships, but maybe you can some how make this relationshop metadata to be expressed in standard way inside the CRD. For example maybe work with openAPI to have a dedicated section of relationships so it will be part of the spec and not just by additional annotations.

3. Consider to add some label-selector into the annotations for CRDs so it will allow your searching to become faster. (from my experience its quit slow at least when i run it on GKE cluster)

4. Show the relationship output as graph, or at least format the relationship output so some existing graph tool can load it and show it nicely.

5. Consider to show service account relationship by default.

6. Looks like you show services automatically, but consider to show also Ingress as well.

7. Consider to add to the README some info about other projects that do something similar with relationships discovery. It will be nice from end user perspective if i can get this analysis also in your page. I know its like show the alternatives for this project but i believe its good way to show the differentiation if exist for this plugin vs the alternatives.

8. Assume the plugin discover objectX which has relationship to objectY but objectY does not exist, then the plugin fail immediately. I believe it should not fail, instead it should continue to show all the relationships and it should state stale relationships in the output but definitely should continue and show the whole picture.

**Potential bug list**:
9. Looks like its not follow PVC -> PV, so consider to add such automatic follow up.

10. Bug or limitation: I encountered issue that I could not annotate CRD with `PodDisruptionBudget` nor `Service` due to error `error: all resources must be specified before annotation changes: PodDisruptionBudget`. Not sure why, but maybe you have limitation that the annotation should be done before creating the first CR from it.

**Comment by devdattakulkarni** [comment]
@shay-berman These are great inputs! Thanks for the suggestions. Some of these are in our roadmap. Will update here as we complete these.

**Comment by shay-berman** [comment]
Sure @devdattakulkarni 
Let me know if there is something new with this ticket.

**Comment by devdattakulkarni** [comment]
> @devdattakulkarni I saw your session in kubecon2020 NA about operators, so I decided to try the plugin on [RedisEnterpriseCluster](https://github.com/RedisLabs/redis-enterprise-k8s-docs).
> 

+1


> Opened this ticket to give you some suggestions on improving the plugin:
> 
> **Suggestions list**:
> 
> 1. Since most of CR creates an app (type of sts, deplp or daemonset) I would suggest you that by design when you run kubectl connections CRDx objecty you should look also on these apps types by default.

While this is probably true for CRs that are used for creating some stateful service (Redis, Cassandra, etc.), it is not generically true. For instance, the NetworkAttachmentDefinition CR in Multus does not create any sub resources. It registers annotations that need to be defined on Pods for them to get multiple IP addresses. Or, there is ClusterIssue CR in CertManager. It also does not create any sub resource.

> 2. Adding annotations is good way to indicate for relationships, but maybe you can some how make this relationshop metadata to be expressed in standard way inside the CRD. For example maybe work with openAPI to have a dedicated section of relationships so it will be part of the spec and not just by additional annotations.

The advantage of using annotations is that relationship information can be added later after a CRD has been registered. With CRD Spec, that is not possible. This is helpful in two ways. Say someone other than the Operator developer needs to add the relationship. Using annotations makes it possible without having to modify the CRD spec/chart. Or, say an Operator's implementation changes and the Operator starts creating different set of sub resources or changes the name of the label selector that it works with. Making such changes to the relationship is easier with CRD annotations than CRD spec. 

> 3. Consider to add some label-selector into the annotations for CRDs so it will allow your searching to become faster. (from my experience its quit slow at least when i run it on GKE cluster)

Yes, improving the discovery performance is on our list.

> 4. Show the relationship output as graph, or at least format the relationship output so some existing graph tool can load it and show it nicely.

Yep. We have been looking into this and will have something to share as soon as it is ready.

> 5. Consider to show service account relationship by default.

Do you mean the Pod <-> Service Account relationship? That is a good suggestion.

> 6. Looks like you show services automatically, but consider to show also Ingress as well.

Sure. We can add that.

> 7. Consider to add to the README some info about other projects that do something similar with relationships discovery. It will be nice from end user perspective if i can get this analysis also in your page. I know its like show the alternatives for this project but i believe its good way to show the differentiation if exist for this plugin vs the alternatives.

So the one project that shows owner relationships is the kubectl-tree plugin. But kubeplus connections goes beyond just the ownerReference relationships. We are able to discover all four kinds of relationships - owners, labels, annotations, spec properties. We haven't seen any others that are able to discover all four kinds of relationships for both built-in and custom resources.

> 8. Assume the plugin discover objectX which has relationship to objectY but objectY does not exist, then the plugin fail immediately. I believe it should not fail, instead it should continue to show all the relationships and it should state stale relationships in the output but definitely should continue and show the whole picture.

This is a good suggestion. We will look into it.

> 
> **Potential bug list**:
> 9. Looks like its not follow PVC -> PV, so consider to add such automatic follow up.

We will look into this.

> 
> 1. Bug or limitation: I encountered issue that I could not annotate CRD with `PodDisruptionBudget` nor `Service` due to error `error: all resources must be specified before annotation changes: PodDisruptionBudget`. Not sure why, but maybe you have limitation that the annotation should be done before creating the first CR from it.

That is indeed strange. We will look into it.
The order of annotating a CRD and CR should not really matter. 


@shay-berman These are all very good suggestions/comments. Really appreciate the feedback. Please keep it coming.

**Comment by devdattakulkarni** [comment]
> > @devdattakulkarni I saw your session in kubecon2020 NA about operators, so I decided to try the plugin on [RedisEnterpriseCluster](https://github.com/RedisLabs/redis-enterprise-k8s-docs).
> 
> +1
> 
> > Opened this ticket to give you some suggestions on improving the plugin:
> > **Suggestions list**:
> > 
> > 1. Since most of CR creates an app (type of sts, deplp or daemonset) I would suggest you that by design when you run kubectl connections CRDx objecty you should look also on these apps types by default.
> 
> While this is probably true for CRs that are used for creating some stateful service (Redis, Cassandra, etc.), it is not generically true. For instance, the NetworkAttachmentDefinition CR in Multus does not create any sub resources. It registers annotations that need to be defined on Pods for them to get multiple IP addresses. Or, there is ClusterIssue CR in CertManager. It also does not create any sub resource.
> 
> > 1. Adding annotations is good way to indicate for relationships, but maybe you can some how make this relationshop metadata to be expressed in standard way inside the CRD. For example maybe work with openAPI to have a dedicated section of relationships so it will be part of the spec and not just by additional annotations.
> 
> The advantage of using annotations is that relationship information can be added later after a CRD has been registered. With CRD Spec, that is not possible. This is helpful in two ways. Say someone other than the Operator developer needs to add the relationship. Using annotations makes it possible without having to modify the CRD spec/chart. Or, say an Operator's implementation changes and the Operator starts creating different set of sub resources or changes the name of the label selector that it works with. Making such changes to the relationship is easier with CRD annotations than CRD spec.
> 
> > 1. Consider to add some label-selector into the annotations for CRDs so it will allow your searching to become faster. (from my experience its quit slow at least when i run it on GKE cluster)
> 
> Yes, improving the discovery performance is on our list.
> 
> > 1. Show the relationship output as graph, or at least format the relationship output so some existing graph tool can load it and show it nicely.
> 
> Yep. We have been looking into this and will have something to share as soon as it is ready.
> 
> > 1. Consider to show service account relationship by default.
> 
> Do you mean the Pod <-> Service Account relationship? That is a good suggestion.
> 
> > 1. Looks like you show services automatically, but consider to show also Ingress as well.
> 
> Sure. We can add that.

@shay-berman Service <-> Ingress is already being discovered. Can you share example of where it did not work for you? If so, that might be a bug.

> 
> > 1. Consider to add to the README some info about other projects that do something similar with relationships discovery. It will be nice from end user perspective if i can get this analysis also in your page. I know its like show the alternatives for this project but i believe its good way to show the differentiation if exist for this plugin vs the alternatives.
> 
> So the one project that shows owner relationships is the kubectl-tree plugin. But kubeplus connections goes beyond just the ownerReference relationships. We are able to discover all four kinds of relationships - owners, labels, annotations, spec properties. We haven't seen any others that are able to discover all four kinds of relationships for both built-in and custom resources.
> 
> > 1. Assume the plugin discover objectX which has relationship to objectY but objectY does not exist, then the plugin fail immediately. I believe it should not fail, instead it should continue to show all the relationships and it should state stale relationships in the output but definitely should continue and show the whole picture.
> 
> This is a good suggestion. We will look into it.
> 
> > **Potential bug list**:
> > 9. Looks like its not follow PVC -> PV, so consider to add such automatic follow up.
> 
> We will look into this.

This is being tracked as part of https://github.com/cloud-ark/kubeplus/issues/624

> 
> > 1. Bug or limitation: I encountered issue that I could not annotate CRD with `PodDisruptionBudget` nor `Service` due to error `error: all resources must be specified before annotation changes: PodDisruptionBudget`. Not sure why, but maybe you have limitation that the annotation should be done before creating the first CR from it.
> 
> That is indeed strange. We will look into it.
> The order of annotating a CRD and CR should not really matter.
> 
> @shay-berman These are all very good suggestions/comments. Really appreciate the feedback. Please keep it coming.

**Comment by devdattakulkarni** [comment]
@shay-berman Support for PersistentVolume and ServiceAccount discovery has now been added. Please pull the latest code/install latest plugins and see if it works for you.

**Comment by devdattakulkarni** [comment]
@shay-berman Here are more updates.

I tried the redislabs example (see https://github.com/cloud-ark/kubeplus/tree/master/examples/redislabs) and here are my observations. I first created the CR instance and then added annotations on the CRD. There was no issue. So this confirms that there is no ordering requirement between when a CRD is annotated and when a CR instance is created.

To address performance issue, we have now added caching for certain calls. There is now approx. 15% improvement in the connections performance. Without caching RedisLabs example was taking approx. 1 minute against my GKE cluster. With caching it is down to approx. 50 seconds.

You can try the latest plugin and let me know how it performs on your side.

**Comment by devdattakulkarni** [comment]
One more update.

We have done a modification such that the resource existence check is happening only at the start and not for every resource that is discovered. This addresses point 8 that you had identified above.

**Comment by devdattakulkarni** [comment]
> @devdattakulkarni I saw your session in kubecon2020 NA about operators, so I decided to try the plugin on [RedisEnterpriseCluster](https://github.com/RedisLabs/redis-enterprise-k8s-docs).
> 
> Opened this ticket to give you some suggestions on improving the plugin:
> 
> **Suggestions list**:
> 
> 1. Since most of CR creates an app (type of sts, deplp or daemonset) I would suggest you that by design when you run kubectl connections CRDx objecty you should look also on these apps types by default.

We have now added this support. Composition annotation is no longer necessary. If it is not specified, we will use a default set of Kinds to search.

> 2. Adding annotations is good way to indicate for relationships, but maybe you can some how make this relationshop metadata to be expressed in standard way inside the CRD. For example maybe work with openAPI to have a dedicated section of relationships so it will be part of the spec and not just by additional annotations.
> 3. Consider to add some label-selector into the annotations for CRDs so it will allow your searching to become faster. (from my experience its quit slow at least when i run it on GKE cluster)

> 4. Show the relationship output as graph, or at least format the relationship output so some existing graph tool can load it and show it nicely.

This has been now added as one of the output types to 'kubectl connections' (the -o png option).

> 5. Consider to show service account relationship by default.

Added.

> 6. Looks like you show services automatically, but consider to show also Ingress as well.

This is already supported.

> 7. Consider to add to the README some info about other projects that do something similar with relationships discovery. It will be nice from end user perspective if i can get this analysis also in your page. I know its like show the alternatives for this project but i believe its good way to show the differentiation if exist for this plugin vs the alternatives.

> 8. Assume the plugin discover objectX which has relationship to objectY but objectY does not exist, then the plugin fail immediately. I believe it should not fail, instead it should continue to show all the relationships and it should state stale relationships in the output but definitely should continue and show the whole picture.

This is now supported.

> 
> **Potential bug list**:
> 9. Looks like its not follow PVC -> PV, so consider to add such automatic follow up.

This is now supported.

> 
> 1. Bug or limitation: I encountered issue that I could not annotate CRD with `PodDisruptionBudget` nor `Service` due to error `error: all resources must be specified before annotation changes: PodDisruptionBudget`. Not sure why, but maybe you have limitation that the annotation should be done before creating the first CR from it.

While we were not able to see this issue, now that we have made the 'resource/composition' annotation optional, you can skip the annotate step. Of course, for accurate sub-resource tracking providing the composition annotation with exact list of sub-resources is necessary. But if the default set used by KubePlus is sufficient then this step is not needed.

**Comment by devdattakulkarni** [comment]
@shay-berman Based on your feedback, we have made several updates to KubePlus. 

You should see much better performance from the 'connections' plugin now. There was one API call that was required only at the beginning, but was happening more times. After fixing this, and also with the additional caching of objects, performance has improved significantly as compared to previous cut of the plugin. 

Overall - really appreciate all the feedback. 

Would love to hear how you find all the improvements.

**Comment by devdattakulkarni** [comment]
Hi @shay-berman We have addressed most of the points reported in this issue.
Whenever you get a chance please give a try and let me know if it addresses your requirements related to Redis Operator.

The documentation is also updated with more details about KubePlus components. It is available here:
https://cloud-ark.github.io/kubeplus/docs/html/html/index.html

**Comment by devdattakulkarni** [comment]
There have been several updates to KubePlus since this issue was reported. Also, all the inputs provided here have been addressed. I am going to close this issue for now. 

kubectl connections plugin has had several improvements. Please try the latest kubectl connections plugins and let us how it looks.

**Comment by shay-berman** [comment]
sure, thanks @devdattakulkarni 
I will try it sooner or later.

**Comment by shay-berman** [comment]
ok quick feedback after you closed https://github.com/cloud-ark/kubeplus/issues/627

I have GKE cluster running kube-prometheus-stack with Prometheus CRD.

According to the issues I mentioned in the description of the issue:
1. works well - now connections on CR will show all the basic resources relationship as well like sts, deploy, pod.
2. Works well - no there is no need for special annotation that the connections will show up. Looks like you did good job to identify the relationship automatically as expected.
3. Not sure - did u added any label-selector, but it is still very slow to run `kubectl connections Pod prometheus-prom-kube-prometheus-stack-prometheus-0 prometheus`  actually can take more then 2 minutes on GKE.
4. still not show how i can show the kubectl connection output as virtual graph. any reference? 
5. Works well - looks like now it show automatically service account relationship.
6. Works well - i see also Ingress showed automaticaly in the connections. well done.
7. Not sure - Any other plugins that do the same relationship thing? if so what makes this project different?
8. Not sure - do u show elements that was deleted but you expect to see them based on relationship?
9. Works well - it now follow PVC -> PV as expected. well done.
10. Not sure about this issue, but maybe u fixed it.

Additional feedback: 
11. Try to use -o pretty as default for -o option.
12. to show graph is the most interesting thing. is it free? how can i show a graph from the output?
13. What the diff between composition to connections?

**Comment by devdattakulkarni** [comment]
@shay-berman wow! this is awesome feedback!! Thanks a LOT. 

About 4 and 12 - Graphical output is free and is available through the png output option (-o png). Try following to get a graphical output:

`kubectl connections Pod prometheus-prom-kube-prometheus-stack-prometheus-0 prometheus -o png`

I will reply to other points separately.

**Comment by devdattakulkarni** [comment]
@shay-berman We have pushed a slew of changes to KubePlus (to both - server-side component and kubectl plugins). One of the updates was to improve the speed of 'kubectl connections'. Do you mind giving it a spin? It will be tremendously helpful! (You will need to download the latest plugins).

From your above list, speed improvement is number 3.

Here are answers/comments about other points.

4 and 12. Graphical output is free and is available through the png output option (-o png). Try following to get a graphical output: `kubectl connections Pod prometheus-prom-kube-prometheus-stack-prometheus-0 prometheus -o png`

7. Other plugins - we know of kubectl-tree. But it only shows owner reference based relationships. As compared to that, KubePlus connections discovers label-based relationships, annotation-based relationships and spec property based relationships, in addition to owner references.

8. Tracking deleted resources - If a resource is not present then it won't get discovered using kubectl connections. kubectl connections will discover all the resources that are still present in the cluster.

10. Yes. It should not be an issue now. (Let us know if you encounter it again).

11. The output option flags for kubectl connections that we currently have are - png, json, flat. png gives graphical output, json is json formattted output, and flat is textual output. Regarding calling one of these as 'pretty', I am not quite sure (or convinced) that any of these will be considered 'pretty'. That is the reason we did not choose the flag name 'pretty'. About the default output option - currently it is 'flat'. The reason for this is - the flat option starts providing immediate feedback as compared to json or png options. The png option is the slowest between the three as it internally first uses json option to generate the json output and then renders the graphical output from the json. Given all these points, I believe that the current set of options and default value strike the right balance between time-to-feedback and understandability. 

13. Difference between composition and connections - This has now become a moot point as we have deprecated composition.

**Comment by devdattakulkarni** [comment]
@shay-berman Just wanted to ping back to inform you that for now I am going to close this issue. We have made several updates to KubePlus over the course of last several months. I would encourage you to give it a spin. Try KubePlus in its entirety, not just the kubectl plugins. Complete documentation of KubePlus is available [here](https://cloud-ark.github.io/kubeplus/docs/html/html/index.html).
Add your comments/thoughts to this issue (or open a new issue).

**Comment by devdattakulkarni** [comment]
@shay-berman Hi!
I just wanted to let you know that we are starting community meetings for KubePlus starting next week. 
We will be meeting every Tuesday at 10.30 a.m. US Central time. Since you had provided such a fantastic feedback earlier, I thought of reaching out to you about the community meetings. I hope you will be able to join us in the meetings and help shape the future of KubePlus.
